Blueprints (CIP-0057) as build artifacts

KtorZ opened the pull request:
Still TODO:

- [x] Add a new `aiken address` to easily compute the address of a given validator.
- [ ] Amend the CIP-0057 PR to reflects changes adopted while implementing this.
- [x] Update the website's user guides 

---
---

- :round_pushpin: **Draft basic Blueprint schema type definition.**
    This doesn't include validations yet. Let's start simple
  and try to get some basic schema generated already.

- :round_pushpin: **Refactor build steps to generate blueprints instead**
    The blueprint is generated at the root of the repository and is
  intended to be versioned with the rest. It acts as a business card
  that contains many practical information. There's a variety of tools
  we can then build on top of open-source contracts. And, quite
  importantly, the blueprint is language-agnostic; it isn't specific to
  Aiken. So it is really meant as an interop format within the
  ecosystem.

- :round_pushpin: **Refactor test collection to use the new CheckedModules method**
    And also, caught a little issue regarding the filtering of test cases.

- :round_pushpin: **Move UPLC dump into separate function + log event.**
    ```
    Compiling aiken-lang/stdlib 43d8e740ffdf5febc59e51b7f0d5f8506115340c (examples/hello_world/build/packages/aiken-lang-stdlib)
    Compiling aiken-lang/hello_world 1.0.0 (examples/hello_world)
   Generating project's blueprint (examples/hello_world/plutus.json)
    Exporting UPLC (examples/hello_world/artifacts)
  ```

- :round_pushpin: **Add title to blueprint's validators**
    And use it to prefix UPLC artifacts' names.

- :round_pushpin: **Add title and description to exported types in the blueprint**
    This also fixes a bug where the documentation of record constructor arguments would be dropped after type-checking. Took me a while to pinpoint.

- :round_pushpin: **Refactor blueprint & handle annotated schemas**
    This also now introduce two levels of representable types (because it's needed at least for tuples):

  Plutus Data (a.k.a Data) and UPLC primitives / constants (a.k.a Schema).

  In practice, we don't want to specify blueprints that use direct UPLC primitives because there's little support for producing those in the ecosystem. So we should aim for producing only Data whenever we can. Yet we don't want to forbid it either in case people know what they're doing. Which means that we need to capture that difference well in the type modelling (in Rust and in the CIP-0057 specification).

  I've also simplified the error type for now, just to provide some degree of feedback while working on this. I'll refine it later with proper errors.

- :round_pushpin: **Allow testing blueprint generation from Aiken programs**
    This is quite something, because now we have a testing pipeline that
  can also be used for testing other compiler-related stuff such as the
  type-checker or the code generator.

- :round_pushpin: **Support tuples in blueprint generation.**
  
- :round_pushpin: **Add support for generics in the blueprint generation.**
  
- :round_pushpin: **Fix nested generics and phantom-types, and handle list special case**
    List of pairs are actually encoded as 'map'.

- :round_pushpin: **Handle opaque single-variant-single-field special case.**
  
- :round_pushpin: **Fix help on unknown::module error.**
  
- :round_pushpin: **Improve error reporting for the blueprint generation.**
    Actually link schema error to source code with a span and a label. This is easily done and provides some extra context.

- :round_pushpin: **Use IndexMap throughout**
     In an ideal world, I should have handlded that directly at the conflicting commit in the rebase, but this would have bubbled up through all commits... which I wasn't really quite keen on going through. So here's an extra ugly commit that comes and 'fix the rebase'.

- :round_pushpin: **Replace 'assets' with generated blueprints.**
  
- :round_pushpin: **Better use of From/Tryfrom within the blueprint module.**
  
- :round_pushpin: **Define serde's Serialize/Deserialize for Program<DeBrujin>**
    This will be useful to re-use this behavior in other structure that contains a Program<DeBruijn> without having to manually serialize or deserialize the entire structure.

- :round_pushpin: **Implement serde's Deserialize for blueprints.**
    Here's a trick though: I got lazy (a bit) and did not write a full deserializer for Schema because this is busywork and not at all necessary at this stage. Instead, I've made the blueprint parameterized by a generic type <T>; which represents the type of the underlying blueprint's schema. When deserializing from JSON, we can default to 'Value' to get a free deserializer. Since all we're interested about is the program and the metadata (purpose and title) of a validator, it works nicely.

  Serialization however expects a Blueprint<Schema>, and most of the functions operates over a Blueprint<Schema> anyway.

- :round_pushpin: **Implement new command: address**
    This calculates a validator's address from validators found in a blueprint. It also provides a convenient way to attach a delegation part to the validator if needs be. The command is meant to provide a nice user experience and works 'out of the box' for projects that have only a single validator. Just call 'aiken address' to get the validator's address.

  Note that the command-line doesn't provide any option to configure the target network. This automatically assumes testnet, and will until we deem the project ready for mainnet. Those brave enough to run an Aiken's program on mainnet will find a way anyway.

---
---

Errors preview:
<img width="712" alt="Screenshot 2023-01-29 at 12 41 32" src="https://user-images.githubusercontent.com/5680256/215326392-c077a6df-0536-4cf7-a87f-a033520a6976.png">
<img width="718" alt="Screenshot 2023-01-29 at 12 52 46" src="https://user-images.githubusercontent.com/5680256/215326397-9d2bf35f-c45b-498b-a45b-0193bdba3a69.png">
<img width="393" alt="Screenshot 2023-01-29 at 12 53 22" src="https://user-images.githubusercontent.com/5680256/215326398-b4f5c7ea-e9cf-4220-9c3e-af64e6a191fd.png">
<img width="717" alt="Screenshot 2023-01-29 at 12 54 10" src="https://user-images.githubusercontent.com/5680256/215326399-2530518a-3bf7-49af-8e4c-c5b614c06b7f.png">
<img width="718" alt="Screenshot 2023-01-29 at 12 54 54" src="https://user-images.githubusercontent.com/5680256/215326401-0ee27b2d-1021-40b3-988e-fa9de293e427.png">
<img width="717" alt="Screenshot 2023-01-29 at 12 56 12" src="https://user-images.githubusercontent.com/5680256/215326403-ebc61163-7e41-4dfb-a214-43d7385b8001.png">

KtorZ commented on crates/aiken-lang/src/tipo/infer.rs:
```diff
@@ -348,7 +348,7 @@ fn infer_definition(
                                             location,
                                             tipo: t.clone(),
-                                            doc: None,
+                                            doc,
```
This one is solely responsible for 3h of debug. 😢

rvcas replied:
Ooff subtle but deadly

KtorZ commented on crates/aiken-project/src/blueprint/validator.rs:
```diff
@@ -0,0 +1,631 @@
+            ),
+        );
+    }
+}
```
Some blueprint examples here, through quite powerful test scenarios ☝️

rvcas replied:
So awesome

KtorZ commented on crates/aiken-project/src/lib.rs:
```diff
@@ -722,7 +535,7 @@ where
                     let mut match_split_dot = match_test.split('.');
 
-                    let match_module = if match_test.contains('.') {
+                    let match_module = if match_test.contains('.') || match_test.contains('/') {
```
Orthogonal, but caught (and fixed) that one while testing some other stuff. The filter on modules wasn't working quite right. For example `-m aiken/list` wouldn't actually yield what you intuitively expect.

rvcas commented on crates/aiken/src/cmd/new.rs:
```diff
@@ -165,7 +165,7 @@ fn gitignore(root: &Path) -> miette::Result<()> {
             r#"
                 # Aiken compilation artifacts
-                assets/
+                artifacts/
```
Cool and then now we can confidently .gitignore this?

KtorZ replied:
Yup.

rvcas commented on crates/aiken-project/src/blueprint/mod.rs:
```diff
@@ -0,0 +1,123 @@
+}
+
+impl Preamble {
+    pub fn from_config(config: &Config) -> Self {
```
This is a good chance to do

```rust
impl From<&Config> for Preamble { }
```

rvcas commented on crates/aiken-project/src/blueprint/validator.rs:
```diff
@@ -0,0 +1,631 @@
+            "mint" => Purpose::Mint,
+            "withdraw" => Purpose::Withdraw,
+            "publish" => Purpose::Publish,
+            unexpected => panic!("Can't turn '{}' into any Purpose", unexpected),
```
Depending on whether on not we actually expect this to happen we could use TryFrom

rvcas commented on crates/aiken-project/src/blueprint/validator.rs:
```diff
@@ -0,0 +1,631 @@
+    use serde_json::{self, json};
+    use std::{collections::HashMap, path::PathBuf};
+
+    // TODO: Possible refactor this out of the module and have it used by `Project`. The idea would
```
Oooo this cool, I like

KtorZ commented:
@rvcas I've addressed your comments and added a new command `aiken address` to fill the gap that this change was introducing:

<img width="677" alt="Screenshot 2023-01-31 at 15 34 47" src="https://user-images.githubusercontent.com/5680256/215789803-c04d92ff-3b33-4bbb-b47b-c292f6cf3430.png">
<img width="458" alt="Screenshot 2023-01-31 at 15 33 54" src="https://user-images.githubusercontent.com/5680256/215789815-e56412b1-261c-4d14-bc56-1f0f80597758.png">
<img width="908" alt="Screenshot 2023-01-31 at 15 34 24" src="https://user-images.githubusercontent.com/5680256/215789808-8366923f-4ed5-43ea-9818-25355aa70d4e.png">

```
// Adding an extra `mint` validator to the file...
```

<img width="823" alt="Screenshot 2023-01-31 at 15 35 56" src="https://user-images.githubusercontent.com/5680256/215789797-19765e2c-00ee-4dc5-bd3c-912c53db2ecd.png">
<img width="456" alt="Screenshot 2023-01-31 at 15 36 07" src="https://user-images.githubusercontent.com/5680256/215789784-29b15754-049d-4876-9463-3ba170da4c07.png">
